### PR TITLE
Add more left join support

### LINKED ISSUE #2672: Linq query failure with left joins
```
IQueryable<ClienteFisicaPlanosBeneficiosPsqDtoEvento> queryable = session?.Query<Pessoa>()
    .LeftJoin(
        session?.Query<ContratoPlanosBeneficios>(),
        pessoa => pessoa,
        contrato => contrato.IdPessoa,
        (pessoa, contrato) => new { Pessoa = pessoa, Contrato = contrato })
    .Select(x =>
        new ClienteFisicaPlanosBeneficiosPsqDtoEvento
        {
            Identificador = x.Pessoa.IdPessoa,
            DependentesNomeCpf = x.Contrato.PesquisaListaDependente,
        });

int result = queryable.Count();
```

```
Query Source could not be identified: ItemName = manyInners, ItemType = System.Collections.Generic.IEnumerable`1[Genesis.OR.ContratoPlanosBeneficios], Expression = join ContratoPlanosBeneficios contrato in value(NHibernate.Linq.NhQueryable`1[Genesis.OR.ContratoPlanosBeneficios]) on [x] equals [contrato].IdPessoa into IEnumerable`1 manyInners
                        
[StackTrace]
   em NHibernate.Linq.QuerySourceNamer.GetName(IQuerySource querySource)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitQuerySourceReferenceExpression(QuerySourceReferenceExpression expression)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitExpression(Expression expression)
   em NHibernate.Linq.Visitors.QueryModelVisitor.VisitNhJoinClause(NhJoinClause joinClause, QueryModel queryModel, Int32 index)
   em Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   em Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   em NHibernate.Linq.Visitors.QueryModelVisitor.Visit()
   em NHibernate.Linq.Visitors.QueryModelVisitor.GenerateHqlQuery(QueryModel queryModel, VisitorParameters parameters, Boolean root, Nullable`1 rootReturnType)
   em NHibernate.Linq.NhLinqExpression.Translate(ISessionFactoryImplementor sessionFactory, Boolean filter)
   em NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
   em NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)
   em NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)
   em NHibernate.Impl.AbstractSessionImpl.CreateQuery(IQueryExpression queryExpression)
   em NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query)
   em NHibernate.Linq.DefaultQueryProvider.Execute(Expression expression)
   em NHibernate.Linq.DefaultQueryProvider.Execute[TResult](Expression expression)
   em Genesis.OR.Pesquisa.ObjetosDto.ClienteFisicaPlanosBeneficiosPsqDtoEvento.ObterQueryPesquisa(IStatelessSession session, Int64 empresaAcessada) na D:\Fontes\Genesis.NET\Genesis.OR\Pesquisa\ObjetosDto\ClienteFisicaPlanosBeneficiosPsqDtoEvento.cs:linha 358
   em Genesis.BO.Pesquisa.ClienteFisicaPlanosBeneficiosPsqBoEvento.Get() na D:\Fontes\Genesis.NET\Genesis.BO\Pesquisa\ClienteFisicaPlanosBeneficiosPsqBoEvento.cs:linha 42
   em lambda_method(Closure , Object , Object[] )
   em Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionMethodAsync>d__12.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextActionFilterAsync>d__10.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   em Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   em Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeInnerFilterAsync>d__14.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResourceFilter>d__22.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   em Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   em Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeFilterPipelineAsync>d__17.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   em Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeAsync>d__15.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   em Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- Fim do rastreamento de pilha do local anterior onde a exceção foi gerada ---
   em System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   em System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   em Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
```

### PR DESCRIPTION
**Pull request by gliljas**
Fixes #2672

LINQ LeftJoin only works if the outer reference is projected in the query result, but it's perfectly feasible to only use it in e.g a where clause.

The new visitor could be merged a bit with the old one, in order to allocate a bit less, but I wanted to keep the old one as is for now, since it's used elsewhere. Perhaps the aggregated group joins should be investigated too.

I added two other failing test cases as well, marked with KnownBug. One which doesn't use the outer reference at all. Not a very likely query, since the join becomes irrelevant, but in query generation scenarios it could happen. I guess a visitor which purges the query model from unused left joins could do the trick. 

The other test case runs a nested join. I think that could be rewritten to chain the nested joins, instead of nesting them.

 Unfortunately, I wasn't able to run the async generator. "Failure scanning NAnt.NUnit2Tasks.dll for extensions"

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
> Unfortunately, I wasn't able to run the async generator. "Failure scanning NAnt.NUnit2Tasks.dll for extensions"

On master branch, this happens if you have kept somewhere a `global.json` file downgrading the .Net Core SDK.
Since this file is required when generating on 5.3.x branches, I just rename it when I am generating on master branch.

**Comment by gliljas** [comment]
Aha, thanks!

**Comment by maca88** [comment]
> The new visitor could be merged a bit with the old one

I updated the new visitor to use an instance of the old one to reduce the allocation.

> I added two other failing test cases as well, marked with KnownBug. One which doesn't use the outer reference at all

I added a fix for the one that doesn't use the outer reference anywhere as it wasn't that complicated.

File: src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs [hunk 1]
```diff
@@ -34,6 +36,7 @@ public static IsAggregatingResults Visit(IEnumerable<GroupJoinClause> groupJoinC
 
 		protected override Expression VisitSubQuery(SubQueryExpression expression)
 		{
+			//Visit the entire query model?
```
Comment (maca88) [inline src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs:39]: I don't see the need to visit the entire query model as aggregated expressions should always be located in the selector.

**Comment by gliljas** [comment]
Thanks, @maca88 !

File: src/NHibernate.Test/Linq/ByMethod/JoinTests.cs [hunk 2]
```diff
@@ -103,6 +99,98 @@ public void LeftJoinExtensionMethodWithMultipleKeyProperties()
...
+							   .Select(x => new { SerialNumber = x.animal.SerialNumber })
+							   .ToList();
+
+				var sql = sqlSpy.GetWholeLog();
+				Assert.That(animals.Count, Is.EqualTo(1));
+				Assert.That(GetTotalOccurrences(sql, "left outer join"), Is.EqualTo(1));
+			}
+		}
+
+		[KnownBug("GH-2379")]
```
Comment (bahusoid) [inline src/NHibernate.Test/Linq/ByMethod/JoinTests.cs:123]: It's not 2379. It's #2739 

And what about https://github.com/nhibernate/nhibernate-core/issues/2672. Is it related/same?

I believe should be reproduced using the following slightly modified test case:
```C#
//GH-2672
[Test]
public void LeftJoinExtensionMethodWithOuterReferenceInWhereClauseOnlyCount()
{
	using (var sqlSpy = new SqlLogSpy())
	{
		var count= db.Animals
						.LeftJoin(
							db.Mammals,
							x => x.Id,
							x => x.Id,
							(animal, mammal) => new {animal, mammal})
						.Where(x => x.mammal.SerialNumber.StartsWith("9"))
						.Select(x => new { x.mammal.Pregnant})
							//.Select(x => new {SerialNumber = x.animal.SerialNumber, x.animal.Description})
						//.ToList()
						.Count();

		var sql = sqlSpy.GetWholeLog();
		Assert.That(count, Is.EqualTo(1));
		Assert.That(GetTotalOccurrences(sql, "left outer join"), Is.EqualTo(1));
	}
}
```

Comment (maca88) [reply src/NHibernate.Test/Linq/ByMethod/JoinTests.cs]: Fixed the bug number and added a fix for #2672.

**Review by maca88 — APPROVED** [review]

File: src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs (hunk 1 above)
Comment (bahusoid) [reply src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs:39]: So shouldn't this comment be removed?

File: src/NHibernate.Test/Linq/ByMethod/JoinTests.cs (hunk 2 above)
Comment (bahusoid) [reply src/NHibernate.Test/Linq/ByMethod/JoinTests.cs]: Fix doesn't look related.. Am I right? Maybe better to fix in separate PR? Just a nit

File: src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs (hunk 1 above)
Comment (maca88) [reply src/NHibernate/Linq/GroupJoin/GroupJoinAggregateDetectionVisitor.cs]: I removed it as I cannot think of any scenario where the whole query model would need to be visited.

**Comment by burhanugur** [comment]
Hi @fredericDelaporte @bahusoid ,

I have the same problem with #2672  . This pull request says it has resolved the issue. When do you plan to merge this request?

**Review by fredericDelaporte — APPROVED** [review]

**Comment by fredericDelaporte** [comment]
For some unknown reason, GitHub seems to have difficulties getting the TeamCity builds statuses. Checking manually on https://teamcity.jetbrains.com/, they have all succeeded for this PR. So, I consider forcing the merge for this PR.

**Comment by 0zancan** [comment]
After the merge i keep having the same issue with #3104.

Nhibernate Version: **5.4.0-dev.3965**

**Comment by bahusoid** [comment]
Most likely fix is not there.. Try with [5.4.0-dev.3980](https://www.myget.org/feed/nhibernate/package/nuget/NHibernate/5.4.0-dev.3980)

**Comment by 0zancan** [comment]
it works with v3980, thank you :)

**Pull request merged**